mitmproxy/addons/dns_resolver.py

```diff
@@ -19,129 +19,41 @@ def __init__(self, response_code: int) -> None:
-            # https://stackoverflow.com/questions/66755681/getaddrinfo-c-on-windows-not-handling-ipv6-correctly-returning-error-code-1
-            raise ResolveError(dns.response_codes.SERVFAIL)  # pragma: no cover
-    return map(
-        lambda addrinfo: dns.ResourceRecord(
-            name=question.name,
-            type=question.type,
-            class_=question.class_,
-            ttl=dns.ResourceRecord.DEFAULT_TTL,
-            data=ip(addrinfo[4][0]).packed,
-        ),
-        addrinfos,
-    )
-
-
-async def resolve_question_by_addr(
-    question: dns.Question,
-    loop: asyncio.AbstractEventLoop,
-    suffix: str,
-    sockaddr: Callable[[list[str]], tuple[str, int] | tuple[str, int, int, int]],
-) -> Iterable[dns.ResourceRecord]:
-    try:
-        addr = sockaddr(question.name[: -len(suffix)].split(".")[::-1])
-    except ValueError:
-        raise ResolveError(dns.response_codes.FORMERR)
-    try:
-        name, _ = await loop.getnameinfo(addr, flags=socket.NI_NAMEREQD)
-    except socket.gaierror as e:
-        raise ResolveError(
-            dns.response_codes.NXDOMAIN
-            if e.errno == socket.EAI_NONAME
-            else dns.response_codes.SERVFAIL
-        )
-    return [
-        dns.ResourceRecord(
-            name=question.name,
-            type=question.type,
-            class_=question.class_,
-            ttl=dns.ResourceRecord.DEFAULT_TTL,
-            data=dns.domain_names.pack(name),
+class DnsResolver:
+    def load(self, loader):
+        loader.add_option(
+            "use_hosts_file",
+            bool,
+            True,
+            "Use the hosts file for DNS lookups in regular DNS mode/wireguard mode.",
         )
-    ]
-
-
-async def resolve_question(
-    question: dns.Question, loop: asyncio.AbstractEventLoop
-) -> Iterable[dns.ResourceRecord]:
-    """Resolve the question into resource record(s), throwing ResolveError if an error condition occurs."""
 
-    if question.class_ != dns.classes.IN:
-        raise ResolveError(dns.response_codes.NOTIMP)
-    if question.type == dns.types.A:
-        return await resolve_question_by_name(
-            question, loop, socket.AddressFamily.AF_INET, ipaddress.IPv4Address
+        loader.add_option(
+            "name_servers",
+            Sequence[str],
+            [],
+            "Name servers to use for lookups. Default: operating system's name servers",
         )
-    elif question.type == dns.types.AAAA:
-        return await resolve_question_by_name(
-            question, loop, socket.AddressFamily.AF_INET6, ipaddress.IPv6Address
+
+    def configure(self, updated):
+        if "use_hosts_file" in updated or "name_servers" in updated:
+            del self.resolver
+            del self.name_servers
+
+    @cached_property
+    def resolver(self) -> mitmproxy_rs.DnsResolver:
+        return mitmproxy_rs.DnsResolver(
+            name_servers=ctx.options.name_servers
+            or mitmproxy_rs.get_system_dns_servers(),
```

Comment:
```suggestion
            name_servers=self.name_servers
```
Adding this change along with the suggested ones